### PR TITLE
Indexer performance upgrade and bugfixes

### LINKED ISSUE #1961: Order parameter for blockchain/box/unspent endpoints
/blockchain​/box​/unspent​/byAddress
and
/blockchain​/box​/unspent​/byErgoTree

APIs need an order parameter to set if the first returned boxes should be the oldest or the newest. At the moment, newest boxes are returned first. For spending boxes, the oldest boxes are the most interesting ones to spent to avoid storage fee.

### PR DESCRIPTION
**Pull request by jellymlg**
- Changed the implementation from using ArrayBuffers to HashMaps. This allows bigger buffer sizes and faster indexing, which made sync times go from ~2,5h to ~1,5h.

- Added a "sortDirection" parameter to /blockchain/box/unspent endpoints
	- Closes #1961

- Fixed a critical bug that caused blocks to be skipped after a rollback occured. Because of this the extra database was probably "corrupted" due to missing boxes and transactions.

- Fixed another critical bug in token tracking that caused overwrites in some tokens information.
	- [x] Require extra indexer resync when updating to newer version

### CODE REVIEW

**Review by kushti — APPROVED** [review]
Approved, but I think more comments are needed, as well as code cleaning and more tests

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala [hunk 1]
```diff
@@ -296,3 +312,13 @@ case class BlockchainApiRoute(readersHolder: ActorRef, ergoSettings: ErgoSetting
   }
 
 }
+
+object SortDirection {
+
+  type Type = Byte
```
[the remark this answers is not in the document]
Comment (jellymlg) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala]: done

**Comment by pragmaxim** [comment]
@jellymlg I noticed that `​/blockchain​/transaction​/byAddress​` at `5.0.7` listed `[]` for an address that explorer listed many transactions.

**Comment by jellymlg** [comment]
> @jellymlg I noticed that `​/blockchain​/transaction​/byAddress​` at `5.0.7` listed `[]` for an address for which explorer listed many transactions.

_Fixed a critical bug that caused blocks to be skipped after a rollback occured. Because of this the extra database was probably "corrupted" due to missing boxes._ - and transactions

**Pull request merged**